### PR TITLE
Add level URL to Dance analytics

### PR DESCRIPTION
**Pull request by fisher-alice**
This is a follow-up to https://github.com/code-dot-org/code-dot-org/pull/55318 which added Amplitude logging for Dance Party validation. @breville [suggested](https://github.com/code-dot-org/code-dot-org/pull/55318#issuecomment-1858635396) that we record the page path on which these events occur, especially for pass/fail events that look identical across levels. 

Although `state` values `nextLevelUrl` and `serverLevelId` already exist, this PR adds `currentScriptLevelUrl`, and then includes this path in the Amplitude event. For pages with '/s/', the URL is assigned. 

Screenshot 2024-01-03 at 11 21 26 AM

For pages with '/level/' `undefined` is assigned. 

Screenshot 2024-01-03 at 11 23 57 AM

Note that `serverLevelId` is already assigned the level id in `pageConstants`.

Screenshot 2024-01-03 at 11 24 36 AM

Now the Amplitude event includes:
- `levelPath`: page URL
- `result`: 'PASSED' or 'FAILED'
- `message`: feedback message key
- `userId`

Consulted with @samantha-code and she confirmed the level URL would be very helpful. I asked if she would want to log only failed events, and she said that 'Passed events can be logged tor (it would be interesting to do a query to see how many students passed without first having failed the level)' so logging both passed and failed validation level events.

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: apps/src/StudioApp.js [hunk 1]
```diff
@@ -3454,6 +3454,7 @@ StudioApp.prototype.setPageConstants = function (config, appSpecificConstants) {
       isK1: config.level.isK1,
       appType: config.app,
       nextLevelUrl: config.nextLevelUrl,
+      thisLevelUrl: config.thisLevelUrl,
```
Comment (breville) [inline apps/src/StudioApp.js]: Maybe `currentLevelUrl`, given the special meaning of `this` in JS?

File: dashboard/app/models/script_level.rb [hunk 2]
```diff
@@ -143,6 +143,16 @@ def has_another_level_to_go_to?
...
   end
 
+  def path_for_user
+    build_script_level_path(this_level)
+  end
+
+  # Return this script level
+  def this_level
+    i = script.script_levels.find_index(self)
+    script.script_levels[i]
```
Comment (bethanyaconnor) [inline dashboard/app/models/script_level.rb]: why is this necessary? Why can't you use `build_script_level_path(self)` above?

File: dashboard/app/models/script_level.rb [hunk 3]
```diff
@@ -143,6 +143,16 @@ def has_another_level_to_go_to?
     end
   end
 
+  def path_for_user
```
Comment (bethanyaconnor) [inline dashboard/app/models/script_level.rb]: why `_for_user` here?

File: dashboard/app/models/script_level.rb (hunk 2 above)
Comment (fisher-alice) [reply dashboard/app/models/script_level.rb]: Yes - thanks for pointing this out!

File: dashboard/app/models/script_level.rb (hunk 3 above)
Comment (fisher-alice) [reply dashboard/app/models/script_level.rb]: Thanks! `path` already defined so this was unnecessary.

**Review by bethanyaconnor — APPROVED** [review]

File: dashboard/app/helpers/levels_helper.rb [hunk 4]
```diff
@@ -255,7 +255,8 @@ def app_options
       view_options(
         lesson_position: @script_level.lesson.absolute_position,
         level_position: @script_level.position,
-        next_level_url: @script_level.next_level_or_redirect_path_for_user(current_user, @lesson)
+        next_level_url: @script_level.next_level_or_redirect_path_for_user(current_user, @lesson),
+        current_level_url: @script_level.path,
```
Comment (sanchitmalhotra126) [inline dashboard/app/helpers/levels_helper.rb]: Because this is only defined for script levels, I'm wondering if it's better to name this `current_script_level_url`? Or maybe at least just leaving a comment in `view_options_helper` - just want to make sure there isn't confusion later around why there isn't a `current_level_url` for a `/levels/` page

**Review by sanchitmalhotra126 — APPROVED** [review]
One small comment but LGTM!

File: dashboard/app/helpers/levels_helper.rb (hunk 4 above)
Comment (fisher-alice) [reply dashboard/app/helpers/levels_helper.rb]: Sure - good idea! Will update.

Comment (fisher-alice) [reply dashboard/app/helpers/levels_helper.rb]: @sanchitmalhotra126 @bethanyaconnor - what do you think of updating `next_level_url` to `next_script_level_url` and `nextLevelUrl` to `nextScriptLevelUrl`?

Comment (fisher-alice) [reply dashboard/app/helpers/levels_helper.rb]: Won't consider including change in this PR but just curious what y'all thought. Thanks!

Comment (sanchitmalhotra126) [reply dashboard/app/helpers/levels_helper.rb]: No strong preference - for me, "next level" implies that we're in a level progression, so the fact that it's only present for script levels makes sense to me. I just felt that "current level" was a little more ambiguous since it could apply to both script level and level pages.
